src/sieve.py

```diff
@@ -13,8 +13,18 @@ def sieve(n: int) -> list[int]:
     """
     assert n > 0
     candidates = list(range(2, n + 1))
-    primes = []
-
-    # FIXME: fill out this bit
-
+    primes = [] 
+    while candidates:
+        i=1
+        length = len(candidates)
+        while i < length:
+            if candidates[i] % candidates[0] == 0:
+                    candidates.remove(candidates[i])
+                    length -= 1
+            else: 
+                i += 1 
```

Comment:
Removing elements from the middle of a list is somewhat slow (for reasons I haven't explained yet), so you would be better off building a whole new list of the remaining candidates compared to eliminating them.

```python
def sieve(n: int) -> list[int]:
    primes = []
    candidates = list(range(2, n+1))
    while candidates:
        p, new_candidates = candidates[0], []
        primes.append(p)
        for cand in candidates:
            if cand % p != 0:
                new_candidates.append(cand)
        candidates = new_candidates
    return primes
```

There is also a little less bookkeeping to keep track of.

You could even use list-comprehension to make it simpler still.

```list
def sieve(n: int) -> list[int]:
    primes = []
    candidates = list(range(2, n+1))
    while candidates:
        primes.append(candidates[0])
        candidates = [cand for cand in candidates if cand % primes[-1] != 0]
    return primes
```